# Introduction and Problem Statement

**Introduction**

Beta-lactam antibiotics are a class of antibiotics that inhibit the synthesis of the bacterial cell wall, leading to cell death. They do this by blocking penicillin-binding proteins (PBPs), which are enzymes that catalyze the cross-linking of peptidoglycan, a major component of the bacterial cell wall. Without PBP activity, peptidoglycan cannot be cross-linked, and the bacterial cell wall becomes weak and susceptible to lysis.

Antimicrobial resistance (AMR) is the ability of microorganisms to withstand the effects of antimicrobial drugs. AMR is a serious global health threat, and it is estimated to cause 1-2 million deaths each year. There are many factors that can contribute to AMR, including overuse and misuse of antibiotics, poor infection control practices, and the spread of resistant bacteria.

Drug repurposing is the identification of new therapeutic uses for existing drugs. It is a promising approach to drug discovery, as it can accelerate the development of new treatments for diseases with unmet medical needs. There are many examples of successful drug repurposing, including thalidomide, sildenafil, and rituximab.

**Problem Statement:**

Antibiotic resistance is a growing global concern, and beta-lactams, a major class of antibiotics, have shown a dramatic decline in efficacy in recent years. This necessitates the development of newer and faster alternatives to beta-lactam resistance. This project aims to broaden the scope of alternative solutions to the global problem of antibiotic resistance.

# Importations

In [ ]:
#imports

import pandas as pd
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from google.colab import data_table

# Data Preprocessings

## Data Importation

We used the chembl database to extract 25 ATP-dependent Clp protease proteolytic (clpp) inhibitors as the positive class. We created a random decoy chemical compound set consisting of 85 compounds from chembl database as the negative class. We then used the mordred toolkit to build molecular descriptors for all 110 chemical compounds in our dataset. This resulted in two files: one containing 1614 molecular decriptors for each of the 25 clpp inhibitors and the other containing 1614 molecular decriptors for each of the 85 random chemical compound.

⏬

In [ ]:
# positive class: ATP-dependent Clp protease proteolytic (clpp) inhibitors

positive = pd.read_csv("/data/des_4.csv") # clpp inhibitors compounds
print("shape before proprocessing",positive.shape)
positive = positive.iloc[:,1:]
positive = positive.head(25)
print("shape after preprocessing", positive.shape)

shape before proprocessing (106, 1614)
shape after preprocessing (25, 1613)


In [ ]:
# negative class: random

negative = pd.read_csv("/data/des_2.csv") #random chemical compounds
print("shape before proprocessing",negative.shape)
negative = negative.iloc[:,1:]
print("shape after preprocessing", negative.shape)

shape before proprocessing (85, 1614)
shape after preprocessing (85, 1613)


## Data Mangling and Cleaning

To prepare the dataset for training a machine learning model, we first added binary labels to the dataset and then concatenated it from both classes and shuffled it. Once we had shuffled the data, we looked for missing values and removed them from the dataset. These data preprocessing steps are important because they help to ensure that the data is in a format that is compatible with the machine learning algorithm and that it does not contain any missing values.

⏬

In [ ]:
# data mangling and cleaning

data_table.enable_dataframe_formatter()
positive['labels'] = 1
negative['labels'] = 0
frames = [positive, negative]
data = pd.concat(frames)
print(data.shape)
data.fillna(method ='pad', inplace=True)
data

(110, 1614)


ABC      ABCGG  nAcid  nBase    SpAbs_A   SpMax_A  SpDiam_A  \
0   22.426819  18.243028      0      0  37.507567  2.290527  4.581054   
1   28.816147  22.557659      0      3  47.134629  2.479550  4.959101   
2   31.607127  23.718932      0      0  52.712937  2.425344  4.850687   
3   28.628869  22.300380      0      0  48.015221  2.458294  4.916588   
4   24.467109  19.634654      0      0  40.628559  2.415312  4.830624   
..        ...        ...    ...    ...        ...       ...       ...   
80  25.315579  20.744317      0      0  37.740036  2.445647  4.880853   
81  23.308764  19.230101      0      0  39.837573  2.519325  5.030429   
82  22.003940  19.119250      0      0  37.102148  2.533727  5.067453   
83  27.728860  21.277320      0      0  43.453727  2.445033  4.880237   
84  22.086719  17.250713      0      0  36.724609  2.389501  4.706817   

       SpAD_A   SpMAD_A   LogEE_A  ...     TSRW10          MW        AMW  \
0   37.507567  1.250252  4.285711  ...  64.616280  409.200156   7.178950   
1   47.134629  1.273909  4.530292  ...  73.702766  536.128328   8.647231   
2   52.712937  1.285681  4.621618  ...  77.780518  573.202872   7.852094   
3   48.015221  1.297709  4.525457  ...  73.439868  518.124288   8.635405   
4   40.628559  1.269642  4.369076  ...  67.427325  453.170510   7.552842   
..        ...       ...       ...  ...        ...         ...        ...   
80  37.740036  1.179376  4.393059  ...  83.930327  460.133395   9.202668   
81  39.837573  1.327919  4.334756  ...  79.489530  430.129550   8.115652   
82  37.102148  1.279384  4.288204  ...  64.782600  396.204907   6.950963   
83  43.453727  1.241535  4.486406  ...  87.145719  512.085399  10.450722   
84  36.724609  1.311593  4.270324  ...  78.820867  413.044630  10.074259   

    WPath  WPol  Zagreb1  Zagreb2   mZagreb1  mZagreb2  labels  
0    2846    39    142.0    156.0  10.888889  6.861111       1  
1    4068    57    192.0    221.0  11.652778  8.069444       1  
2    5954    59    206.0    233.0  12.562500  9.111111       1  
3    4202    55    188.0    215.0  10.951389  8.222222       1  
4    2998    44    158.0    177.0   9.979167  7.166667       1  
..    ...   ...      ...      ...        ...       ...     ...  
80   3310    48    174.0    201.0  13.826389  6.472222       0  
81   2326    50    158.0    189.0   8.590278  6.750000       0  
82   2106    48    148.0    176.0   9.840278  6.666667       0  
83   4426    51    188.0    217.0  12.375000  7.305556       0  
84   2376    40    148.0    173.0   8.000000  6.111111       0  

[110 rows x 1614 columns]

## Features Extraction

we used the Pandas library to extract a sub-DataFrame containing the features from our dataset. The features in our dataset consisted of 1613 molecular descriptors spanning over the samples, which came from two major sources.

⏬

In [ ]:
# features extraction
features = data.iloc[:, :-1]

## Labels Extraction

we used the Pandas library to extract a sub-DataFrame containing the labels from our dataset. The labels in our dataset consisted of two classes (1 or 0) spanning over the features of 1613 molecular descriptors.

⏬

In [ ]:
# labels extraction

labels = data.iloc[:,-1:]

## Splitting Dataset

The features and labels parameters are the input and output variables of the dataset, respectively. The test_size parameter specifies the proportion of the dataset to be used for the test set. The random_state parameter is used to ensure that the split is reproducible.

The train_test_split() function returns four variables:

* X_train: The training data features
* X_test: The test data features
* y_train: The training data labels
* y_test: The test data labels

The test_size parameter is set to 0.3, which means that 30% of the dataset will be used for the test set. The random_state parameter is set to 50, which ensures that the split is reproducible.

⏬

In [ ]:
# splitting the dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(features, labels, test_size=0.33, random_state=50)

# Training and Evaluation

## Training

HistGradientBoostingClassifier is a machine learning algorithm that is well-suited for classification tasks. It is an ensemble learning algorithm, which means that it combines the predictions of multiple weak learners to produce a more accurate prediction. The fact that the model is able to achieve 97% accuracy on the held-out test set indicates that the model is not overfitting the training data. Overfitting occurs when the model learns the training data too well and is unable to generalize to new data.

⏬


In [ ]:
from sklearn.ensemble import HistGradientBoostingClassifier
scaler = StandardScaler()
hgbc = HistGradientBoostingClassifier()
pipe_hgbc = Pipeline(steps=[("scaler", scaler), ("hgbc", hgbc)])
pipe_hgbc.fit(X_train, y_train)
print ("Accuracy Score of HistGradientBoostingClassifier", pipe_hgbc.score(X_test, y_test))

/usr/local/lib/python3.10/dist-packages/sklearn/utils/extmath.py:1047: RuntimeWarning: invalid value encountered in divide
  updated_mean = (last_sum + new_sum) / updated_sample_count
/usr/local/lib/python3.10/dist-packages/sklearn/utils/extmath.py:1052: RuntimeWarning: invalid value encountered in divide
  T = new_sum / new_sample_count
/usr/local/lib/python3.10/dist-packages/sklearn/utils/extmath.py:1072: RuntimeWarning: invalid value encountered in divide
  new_unnormalized_variance -= correction**2 / new_sample_count
/usr/local/lib/python3.10/dist-packages/sklearn/utils/validation.py:1143: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Accuracy Score of HistGradientBoostingClassifier 0.972972972972973


## Evaluation Matrix

The precision, recall, and F1 scores for the target class "ATP-dependent Clp protease proteolytic (clpp) inhibitors" are all very high, indicating that the machine learning model is performing very well. A precision of 100% means that all of the predicted positive samples are actually positive. A recall of 96% means that the model is able to identify 96% of the actual positive samples. A F1 score of 98% means that the model has a good balance between precision and recall.

⏬

In [ ]:
#evaluations
from sklearn.metrics import classification_report
predictions_eval = pipe_hgbc.predict(X_test)
target_names = ['1', '0']
print(classification_report(y_test, predictions_eval, target_names=target_names))

              precision    recall  f1-score   support

           1       1.00      0.96      0.98        28
           0       0.90      1.00      0.95         9

    accuracy                           0.97        37
   macro avg       0.95      0.98      0.96        37
weighted avg       0.98      0.97      0.97        37



# Screening of Infectious FDA Approves

We extracted 297 FDA approved drugs from chembl database and preprocessed the data according to the protocol of this project. We then screened the dataset for repurposing clpp inhibitors and identified 2 antiviral drugs.


1. Oseltamivir.
2. Valganciclovir.

Oseltamivir (trade name Tamiflu) is a neuraminidase inhibitor that is used to treat and prevent influenza A and B virus infections. It works by blocking the neuraminidase enzyme, which the virus needs to spread from one cell to another.

Valganciclovir (trade name Valcyte) is a prodrug of ganciclovir, an antiviral drug that is used to treat and prevent cytomegalovirus (CMV) infections. Valganciclovir is converted to ganciclovir in the body after it is taken orally.

Both oseltamivir and valganciclovir are generally safe and well-tolerated drugs. However, they can cause some side effects, such as nausea, vomiting, diarrhea, and headache.

⏬





In [ ]:
# screening
screening_data = pd.read_csv("/data/des_3.csv") # infectious fda apprroved almost 300 compounds
print(screening_data.shape)
screening_data = screening_data.iloc[:,1:]
print("shape after preprocessing", screening_data.shape)

(297, 1614)
shape after preprocessing (297, 1613)


In [ ]:
screening_data.fillna(method ='pad', inplace=True)
screening_data.shape

(297, 1613)

In [ ]:
predictions = pipe_hgbc.predict(screening_data)
predictions = pd.DataFrame(predictions)
predictions

0
0    1
1    1
2    1
3    1
4    1
..  ..
292  0
293  0
294  0
295  0
296  0

[297 rows x 1 columns]

# Conclution and Prospects

Oseltamivir and Valganciclovir are antiviral drugs that have the potential to inhibit the clpp enzyme and act as an alternative to beta-lactam antibiotics. We used the chembl database to extract 25 ATP-dependent Clp protease proteolytic (clpp) inhibitors as the positive class. We created a random decoy chemical compound set consisting of 85 compounds from chembl database as the negative class. HistGradientBoostingClassifier is trained on both the classes with the following evaluations:
> 1. Accuracy = 97%
> 2. Precision = 100%
> 3. Recall = 96%
> 4. F1 Score = 98%

The 297 FDA approved infectious drugs were screened to repurposed 2 antiviral drugs including `Oseltamivir and Valganciclovir as potential and safe drug alternatives for beta-lactams.` These antiviral drugs could act against the global problem of antimicrobial resistence. They need to be further tested for thier activity against the clpp inhibition.

# About Me

I am a Bioinformatics and Machine learning expert, practecing insilico development and analytics in the domain of biology and medicine. I am open for research collaborations, you can email me to discuss.

Thank you,
```
Ahsan Raza

Masters in Bioinformatics

aysanraza@gmail.com

Islamabad, Pakistan.
```

# 💻 Tech Stack:
![Python](https://img.shields.io/badge/python-3670A0?style=for-the-badge&logo=python&logoColor=ffdd54) ![Shell Script](https://img.shields.io/badge/shell_script-%23121011.svg?style=for-the-badge&logo=gnu-bash&logoColor=white) ![Anaconda](https://img.shields.io/badge/Anaconda-%2344A833.svg?style=for-the-badge&logo=anaconda&logoColor=white) ![GitHub](https://img.shields.io/badge/GitHub-%23121011.svg?style=for-the-badge&logo=github&logoColor=white) 	![Neo4J](https://img.shields.io/badge/Neo4j-008CC1?style=for-the-badge&logo=neo4j&logoColor=white) ![SQLite](https://img.shields.io/badge/sqlite-%2307405e.svg?style=for-the-badge&logo=sqlite&logoColor=white) ![MySQL](https://img.shields.io/badge/mysql-%2300f.svg?style=for-the-badge&logo=mysql&logoColor=white) ![scikit-learn](https://img.shields.io/badge/scikit--learn-%23F7931E.svg?style=for-the-badge&logo=scikit-learn&logoColor=white) ![SciPy](https://img.shields.io/badge/SciPy-%230C55A5.svg?style=for-the-badge&logo=scipy&logoColor=%white) ![Plotly](https://img.shields.io/badge/Plotly-%233F4F75.svg?style=for-the-badge&logo=plotly&logoColor=white) ![TensorFlow](https://img.shields.io/badge/TensorFlow-%23FF6F00.svg?style=for-the-badge&logo=TensorFlow&logoColor=white) ![Pandas](https://img.shields.io/badge/pandas-%23150458.svg?style=for-the-badge&logo=pandas&logoColor=white) ![NumPy](https://img.shields.io/badge/numpy-%23013243.svg?style=for-the-badge&logo=numpy&logoColor=white) ![Keras](https://img.shields.io/badge/Keras-%23D00000.svg?style=for-the-badge&logo=Keras&logoColor=white) ![GIT](https://img.shields.io/badge/Git-fc6d26?style=for-the-badge&logo=git&logoColor=white) ![LINUX](https://img.shields.io/badge/Linux-FCC624?style=for-the-badge&logo=linux&logoColor=black)

 ## 💰 You can help me by Donating
  [![BuyMeACoffee](https://img.shields.io/badge/Buy%20Me%20a%20Coffee-ffdd00?style=for-the-badge&logo=buy-me-a-coffee&logoColor=black)](https://buymeacoffee.com/https://www.buymeacoffee.com/aysanraza)